Experimental anyhow integration

njgheorghita opened the pull request:
This is an experimental pr to see what using `anyhow` would look like. We've been talking about this in #214, but I'm not able to truly wrap my head around it without giving it a go. To be fair, this is my first time using the library, and It's entirely possible that my integration is not the cleanest / idiomatic way to go about using this library. If so, please let me know what I'm missing b/c that for sure influences my opinion...

The pros...
- It's cleaner and neater and shorter
- I'm sure it would help boost my speed in rust, not having to worry about custom error types all the time
- The `context` method seems pretty useful for adding detail to messages

The cons...
- It makes me feel nervous... all these untyped errors. smh what has rust done to me
- It seems far too easy to write insufficient error messages. At minimum, using `thiserror` enforces a well thought out message and heirarchy of error types.

Ultimately....
[This blog](https://nick.groenen.me/posts/rust-error-handling/#libraries-versus-applications) inspired the issue. The key advice being `"use thiserror for libraries and anyhow for applications"`. After playing with `anyhow` I'm starting to come to the belief that `trin` isn't really an application. It's closer to a collection of libraries acting as a single application. A well-thought out, consistent, typed error system is something we would benefit from. Especially dealing with bugs on various networks. 

Maybe we use `thiserror` in certain parts of the codebase and `anyhow` in others? Imo, one error library is better than two.

I'm certainly open to having my mind changed. Maybe I've missed out some of the more useful features of `anyhow`. Since this decision touches the entire codebase, I'm curious to hear what everyone thinks....

ogenev commented:
I'm less or more biased towards `anyhow` (because I proposed this :rofl: ) but let me think about this again.

At this stage of the project, I feel iteration speed is more valuable than a well-defined error type system because there are still a lot of unknown implementation details.

It is true that we can consider most of the workspace creates libraries, but as long as we are the only ones that have ownership over those libraries, I think it is pretty straightforward to manage our `anyhow` error handling in a way to helps us with debugging and tracing.
If there is a library that is used by third parties, a good error type structure with `thiserror` makes sense because the application code is responsible for handling those errors and it will need error details.

I'm not concerned about using two error handling libraries, as long as there is only one in any workspace crate. But maybe we are still in the process to define what sub-crates we want to have, for example probably `discovery`, `overlay`, `utp`, and  `json-rpc` should be in separate crates...

mrferris commented on trin-core/src/portalnet/overlay.rs:
```diff
@@ -137,16 +136,22 @@ impl OverlayProtocol {
         &self,
         talk_request: &TalkRequest,
-    ) -> Result<Response, OverlayRequestError> {
+    ) -> anyhow::Result<Response> {
```
What do you think of putting at the top of the file:`type Result = anyhow::Result` so that all of these lines can just be: 
```suggestion
    ) -> Result<Response> {
```

Might improve readability a bit.

njgheorghita replied:
I'm not so keen on this. Especially if we settle on two different error libraries. Yeah, it's a definite improvement in readability, but...
1 - it might confuse people new to codebase, familiar with Rust but not `anyhow`
2 - keeping track of which error handling library is used where in the codebase, I'd get frustrated with having to scroll to the top of each file to see if this import is there or not.

mrferris reviewed (approved):
This is great! Yea, seeing it used definitely shows how much boilerplate/time it saves when all we'll do with the errors is log them out anyway. Great suggestion @ogenev.

I think that using both is fine, I think for example that using `thiserror` in `PortalStorage` is still preferable, because with some of the errors returned we might want to re-try the operation instead of only logging it. I like the idea of having storage, utp, etc broken down into crates with defined error models, while the higher level stuff just emits strings for logs via anywhow.

njgheorghita commented:
Thanks for the thoughts. I'm shifting towards being ok with using both `thiserror` and `anyhow` together. Prs like #242 really help prove the point that sometimes fast > precise when it comes to error handling. But, there are still certain situations where typed errors are fairly crucial (aka in `PortalStorage` like @mrferris mentioned or in the [overlay networks](https://github.com/ethereum/trin/blob/master/trin-history/src/network.rs#L58). 

I guess my rule of thumb would be something like: Use `anyhow` unless there is a concrete benefit to using typed errors. (aka we should never downcast an `anyhow` error to a string and match on that string). 

The next open question to me seems like: how do we want to handle the migration? Should we merge this pr? This pr only touches part of the codebase... Should I extend it to the entire codebase where the rule of thumb applies? Should we close this pr and just gradually migrate over to `anyhow` where applicable in the following prs?

njgheorghita commented on trin-core/src/portalnet/types/uint.rs:
```diff
@@ -139,17 +132,18 @@ mod test {
+        let result: anyhow::Result<U256> = input.try_into();
+        assert_eq!(
+            result.unwrap_err().to_string(),
+            "Failed to convert U512 to U256 due to overflow"
```
I'm a big fan of matching error messages in tests, that way you're guaranteed that the test failed for the exact reason it's was designed to. If somebody wants to change the error message (which shouldn't happen too frequently) then the test will fail with a fairly explicit reason and the fix will be obvious.

ogenev reviewed (approved):
> The next open question to me seems like: how do we want to handle the migration? Should we merge this pr? 

I'm ok with merging this and gradually migrating our errors to anyhow where it is possible in later PRs. I think the early stage of development is the best time to experiment with such things :)

ogenev commented on trin-state/src/network.rs:
```diff
@@ -64,41 +64,13 @@ impl StateNetwork {
+                // Tbh I'm a bit stumped on how to handle this area, so this is just a temporary
+                // solution to get the compiler passing. Will revisit if we decide to continue with
+                // anyhow
+                Err(msg) => {
```
Yeah, I was stumped also here when trying prototyping it and that's why I assumed that getting rid of `OverlayRequestError` may not be so trivial.

njgheorghita commented:
Sounds good. I reverted all of the `OverlayRequestError` -> `anyhow` conversions here, since we couldn't make the transformation 100%, picking and choosing some arbitrary error border (not to mention somewhat clunky conversion b/w error types) for where to use one crate or the other felt messy. I also updated the contributing guidelines to contain both error libraries.
